Source/WebCore/rendering/RenderObject.cpp

```diff
@@ -2282,6 +2282,11 @@ static Vector<FloatRect> absoluteRectsForRangeInText(const SimpleRange& range, T
         return { };
 
     auto offsetRange = characterDataOffsetRange(range, node);
+    // Move to surrogate pair start for Range start and past surrogate pair end for Range end in case the trailing surrogate is indexed.
+    if (U16_IS_TRAIL(node.data()[offsetRange.start]) && offsetRange.start > 0 && U16_IS_LEAD(node.data()[offsetRange.start - 1]))
+        offsetRange.start--;
+    if (U16_IS_TRAIL(node.data()[offsetRange.end]) && offsetRange.end > 0 && U16_IS_LEAD(node.data()[offsetRange.end - 1]))
```

Comment:
Ditto.

Reply:
How does this make sense? Should this be checking U16_IS_LEAD on offsetRange.end and U16_IS_TRAIL on offsetRange.end + 1?

Reply:
I think that is the correct behaviour when taking into account what the spec text says:
"If the range covers a partial [typographic character unit](https://drafts.csswg.org/css-text-4/#typographic-character-unit) (e.g. half a surrogate pair or part of a grapheme cluster), the full typographic character unit must be included for the purpose of computing the bounds of the relevant [DOMRect](https://drafts.fxtf.org/geometry-1/#domrect)." So the complete surrogate pair would need to be included both at the start and end of the range.

Reply:
I get that the desired effect. What I don't think right is the code in PR:
```
U16_IS_TRAIL(node.data()[offsetRange.end]) && offsetRange.end && U16_IS_LEAD(node.data()[offsetRange.end - 1])
```

Oh right, `node.data()[offsetRange.end]` grabs the character after the end, not the one before it. But I think we still need to also check that `offsetRange.end` is less than node.date()'s length. Otherwise, we can end up reading past the end of the buffer.

Reply:
I checked, an exception is thrown when a start/end offset is set on the DOM level. However, for some added security I added two ASSERTS checking both start and end offset values.